kv/dashboard.go

```diff
@@ -0,0 +1,960 @@
+func (s *Service) FindDashboard(ctx context.Context, filter influxdb.DashboardFilter, opts ...influxdb.FindOptions) (*influxdb.Dashboard, error) {
+	if len(filter.IDs) == 1 {
+		return s.FindDashboardByID(ctx, *filter.IDs[0])
+	}
+
+	var d *influxdb.Dashboard
+	err := s.kv.View(func(tx Tx) error {
+		filterFn := filterDashboardsFn(filter)
+		return s.forEachDashboard(ctx, tx, opts[0].Descending, func(dash *influxdb.Dashboard) bool {
+			if filterFn(dash) {
+				d = dash
+				return false
+			}
+			return true
+		})
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if d == nil {
+		return nil, &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  influxdb.ErrDashboardNotFound,
+		}
+	}
+
+	return d, nil
+}
+
+func filterDashboardsFn(filter influxdb.DashboardFilter) func(d *influxdb.Dashboard) bool {
+	if len(filter.IDs) > 0 {
+		m := map[string]struct{}{}
+		for _, id := range filter.IDs {
+			m[id.String()] = struct{}{}
+		}
+		return func(d *influxdb.Dashboard) bool {
+			_, ok := m[d.ID.String()]
+			return ok
+		}
+	}
+
+	return func(d *influxdb.Dashboard) bool { return true }
+}
+
+// FindDashboards retrives all dashboards that match an arbitrary dashboard filter.
+func (s *Service) FindDashboards(ctx context.Context, filter influxdb.DashboardFilter, opts influxdb.FindOptions) ([]*influxdb.Dashboard, int, error) {
+	ds := []*influxdb.Dashboard{}
+	if len(filter.IDs) == 1 {
+		d, err := s.FindDashboardByID(ctx, *filter.IDs[0])
+		if err != nil && influxdb.ErrorCode(err) != influxdb.ENotFound {
+			return ds, 0, &influxdb.Error{
+				Err: err,
+			}
+		}
+		if d == nil {
+			return ds, 0, nil
+		}
+		return []*influxdb.Dashboard{d}, 1, nil
+	}
+	err := s.kv.View(func(tx Tx) error {
+		dashs, err := s.findDashboards(ctx, tx, filter, opts)
+		if err != nil && influxdb.ErrorCode(err) != influxdb.ENotFound {
+			return err
+		}
+		ds = dashs
+		return nil
+	})
+
+	if err != nil {
+		return nil, 0, &influxdb.Error{
+			Err: err,
+		}
+	}
+
+	influxdb.SortDashboards(opts, ds)
+
+	return ds, len(ds), nil
+}
+
+func (s *Service) findOrganizationDashboards(ctx context.Context, tx Tx, orgID influxdb.ID) ([]*influxdb.Dashboard, error) {
+	idx, err := tx.Bucket(orgDashboardIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(desa): support find options.
+	cur, err := idx.Cursor()
+	if err != nil {
+		return nil, err
+	}
+
+	prefix, err := orgID.Encode()
+	if err != nil {
+		return nil, err
+	}
+
+	ds := []*influxdb.Dashboard{}
+	for k, _ := cur.Seek(prefix); bytes.HasPrefix(k, prefix); k, _ = cur.Next() {
+		_, id, err := decodeOrgDashboardIndexKey(k)
+		if err != nil {
+			return nil, err
+		}
+
+		d, err := s.findDashboardByID(ctx, tx, id)
+		if err != nil {
+			return nil, err
+		}
+
+		ds = append(ds, d)
+	}
+
+	return ds, nil
+}
+
+func decodeOrgDashboardIndexKey(indexKey []byte) (orgID influxdb.ID, dashID influxdb.ID, err error) {
+	if len(indexKey) != 2*influxdb.IDLength {
```

Comment:
I wonder if all these errors in decoding should be internal errors.

Reply:
They definitely should be